cloud/services/securitygroups/securitygroups.go

```diff
@@ -59,52 +66,58 @@ func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
 		return errors.New("invalid security groups specification")
 	}
 
-	securityRules := &[]network.SecurityRule{}
+	securityGroup, err := s.Client.Get(ctx, s.Scope.ResourceGroup(), nsgSpec.Name)
+	if err != nil && !azure.ResourceNotFound(err) {
+		return err
+	}
+
+	nsgExists := false
+	var etag string
+	securityRules := make([]network.SecurityRule, 0)
+	if securityGroup.Name != nil {
+		nsgExists = true
+		etag = *securityGroup.Etag
+		securityRules = *securityGroup.SecurityRules
+	}
+
+	sshRule := controlPlaneDefaultRule{"allow_ssh", 100, "22"}
+	apiServerRule := controlPlaneDefaultRule{"allow_6443", 101, strconv.Itoa(int(s.Scope.APIServerPort()))}
 
 	if nsgSpec.IsControlPlane {
-		klog.V(2).Infof("using additional rules for control plane %s", nsgSpec.Name)
-		securityRules = &[]network.SecurityRule{
-			{
-				Name: to.StringPtr("allow_ssh"),
-				SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
-					Protocol:                 network.SecurityRuleProtocolTCP,
-					SourceAddressPrefix:      to.StringPtr("*"),
-					SourcePortRange:          to.StringPtr("*"),
-					DestinationAddressPrefix: to.StringPtr("*"),
-					DestinationPortRange:     to.StringPtr("22"),
-					Access:                   network.SecurityRuleAccessAllow,
-					Direction:                network.SecurityRuleDirectionInbound,
-					Priority:                 to.Int32Ptr(100),
-				},
-			},
-			{
-				Name: to.StringPtr("allow_6443"),
-				SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
-					Protocol:                 network.SecurityRuleProtocolTCP,
-					SourceAddressPrefix:      to.StringPtr("*"),
-					SourcePortRange:          to.StringPtr("*"),
-					DestinationAddressPrefix: to.StringPtr("*"),
-					DestinationPortRange:     to.StringPtr(strconv.Itoa(int(s.Scope.APIServerPort()))),
-					Access:                   network.SecurityRuleAccessAllow,
-					Direction:                network.SecurityRuleDirectionInbound,
-					Priority:                 to.Int32Ptr(101),
-				},
-			},
+		if nsgExists {
+			// Check if the expected rules are present
+			update := false
+			if !ruleExists(securityRules, sshRule) {
```

Comment:
maybe use a for loop over rule array? just for later on if there are more rules to be added.